### PR TITLE
Added packages n98-magerun and n98-magerun2

### PR DESCRIPTION
**Pull request by fballiano**
###### Description of changes

Being a magento developer, n98-magerun and n98-magerun2 are very important tools to have.
The website of the project is https://magerun.net

This is the first package I work on and I'm fairly new to nix thus I don't know how to test the package before it's actually merged in the main repo.

###### Things done

Copied the composer package and modified it for the new tools.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by drupol** [comment]
Hi,

Thanks for improving the PHP ecosystem on Nix !

I've got some questions:

1. Can you explain why 2 versions are required? Usually we try to commit the latest stable version of a project.
2. I would suggest to use [the Github repository](https://github.com/netz98/n98-magerun2) and build the phar from the source ([check how php-parallel-lint is done](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/php-packages/php-parallel-lint/default.nix)). Or you could ask the maintainer of the Github repository to add the phar asset to each release, it could even be automatically done with Github action. The benefit of using Github repository in the src, is the there is a robot([r-ryantm](https://github.com/r-ryantm)) watching the `nixpkgs` repo and create pull request as soon as an update of a package is published.

**Comment by fballiano** [comment]
Hi @drupol and thanks for the quick check!

1. magerun is for Magento1/OpenMage while magerun2 is for Magento2, they're completely different softwares and totally incompatible between each other
2. I checked the github's release page but the phar files are not uploaded there: https://github.com/netz98/n98-magerun2/releases that's why I couldn't use them :-(

**Comment by drupol** [comment]
Thanks!

Do you think we could build the `phar` from the `src`? How about suggesting @netz98 in an issue to upload it at each release?

**Comment by fballiano** [comment]
I can contact @netz98 but I guess it will take time, if they want to do it, and anyway we wouldn't have control if they don't do it in the future.

I'll check the building from source thing although I've never done it.

**Comment by drupol** [comment]
Just to make my point here.

I'm asking you to check if it would be possible to build the phar from the source because security wise, I'm a bit reluctant to use a random phar file and add it in Nix.

Of course, for someone using Magento every day, this is a nonsense because you use the tool everyday and you know that site, but for someone like me who never used anything from Magento... I'm a bit reluctant putting trust into something I never heard of yet.

This is the reason why I prefer to check if it would be possible to build it from the Github source and/or download the asset from Github directly.

**Comment by fballiano** [comment]
I agree with your POV, what I can tell is that the domain used to download the phar is owned by the same owners of the github repo and linked in their repo description.

I'll try to convert these 2 packages to building them from source although it wasn't really my plan since I'm short with time.

**Comment by fballiano** [comment]
seeing the updated first answer I'll close this and see if I can redo it another time.

**Comment by domenkozar** [comment]
@drupol could we allow binary packages until upstream releases the source? I'm also not happy but 

This and a few other things are crucial to support Magento devs with https://devenv.sh/

**Comment by drupol** [comment]
Yes yes we could and we already do, but if we could have things directly from Github and built from the source, it would be nice. Right now, nothing garanties me that the build on the website comes from the sources on Github.

Therefore I would recommend to re-open this PR, move the files in `misc/tools` and wait for the review of the other PHP maintainers.

My goal here is not to block contributions, on the contrary, I'm glad when I see them. However, since I know nothing about Magento, I think it's legitimate to see if there are better alternatives.

I (and other contributors) are here to help and to make Nix great, and I think it's our duty to make sure that new packages are carefully checked.

Anyway, after giving a second thought to it, copying the `phar` is our best option since we can't run `composer` during the `buildPhase`... since we don't have network access.

So, to summarize, things that are left to do:

1. Reopen this PR
2. Move the newly nix files in `misc/tools` 
3. One single commit per derivation (so 2 in total)

Once this is done, I guess other maintainers will comment as well, today is a busy day for me, I'll revisit this thread tomorrow.

**Comment by ktomk** [comment]
If I could get some guidance on how to package for a PHP project, I could pick Magerun 1 as sandbag and check the phar from source build there. I'd also like to add a shell.nix file into sources so that it can be run in shell right from the source tree to have an early intergration of what would become the next build right from the working tree.

**Comment by drupol** [comment]
> If I could get some guidance on how to package for a PHP project, I could pick Magerun 1 as sandbag and check the phar from source build there. I'd also like to add a shell.nix file into sources so that it can be run in shell right from the source tree to have an early intergration of what would become the next build right from the working tree.

Tomorrow I'll be available for a meeting if needed!

**Comment by ktomk** [comment]
> Tomorrow I'll be available for a meeting if needed!

That would be formidable.
